client/components/tinymce/plugins/simple-payments/dialog/index.jsx

```diff
@@ -32,28 +36,162 @@ class SimplePaymentsDialog extends Component {
+			isSubmitting: false,
+			errorMessage: null,
+		};
+	}
+
+	componentDidMount() {
+		this._isMounted = true;
+	}
+
+	componentWillUnmount() {
+		this._isMounted = false;
+	}
+
+	handleFormFieldChange = ( name, value ) => {
+		this.formStateController.handleFieldChange( { name, value } );
+	};
+
+	isFormFieldInvalid = name => formState.isFieldInvalid( this.state.form, name );
+
+	getFormValues() {
+		return formState.getAllFieldValues( this.state.form );
+	}
+
+	validateFormFields( formValues, onComplete ) {
+		const formErrors = {};
+
+		if ( ! formValues.title ) {
+			formErrors.title = [ 'empty' ];
+		}
+
+		if ( ! formValues.price ) {
+			formErrors.price = [ 'empty' ];
+		}
+
+		if ( ! formValues.email ) {
+			formErrors.email = [ 'empty' ];
+		} else if ( ! emailValidator.validate( formValues.email ) ) {
+			formErrors.email = [ 'invalid' ];
+		}
+
+		onComplete( null, formErrors );
+	}
+
 	handleSelectedChange = selectedPaymentId => {
 		this.setState( { selectedPaymentId } );
 	};
 
-	handleInsert = () => {
-		this.props.onInsert( { id: this.state.selectedPaymentId } );
+	handleClose = () => {
+		// clear the form after a successful submit -- it'll be blank next time it's opened
+		this.formStateController.resetFields( this.constructor.initialFields );
+		this.props.onClose();
 	};
 
+	dismissError = () => {
+		this.setState( { errorMessage: null } );
+	};
+
+	handleInsert = async () => {
+		this.setState( { isSubmitting: true } );
+
+		try {
+			const insertedProductId = await this.getInsertedProductId();
+			if ( insertedProductId ) {
+				this.props.onInsert( { id: insertedProductId } );
+				// clear the form after a successful submit -- it'll be blank next time it's opened
+				this.formStateController.resetFields( this.constructor.initialFields );
+			}
+		} catch ( error ) {
+			if ( this._isMounted ) {
+				this.setState( {
+					errorMessage: this.props.translate( 'The payment button could not be inserted.' ),
+				} );
+			}
+		}
+
+		this._isMounted && this.setState( { isSubmitting: false } );
+	};
+
+	/*
+	 * Asynchronously retrive the ID of the product to insert. In case of selection from
+	 * list of existing product, we know the ID right away. When inserting from a filled
+	 * form, we first need to issue a create request to the server, and wait for it to
+	 * return an ID.
+	 * @returns the ID, or null in case the form is not valid, or throws an exception in
+	 * case of error that should be displayed as a notice.
+	 */
+	async getInsertedProductId() {
+		const { activeTab, siteId, dispatch } = this.props;
+
+		if ( activeTab === 'paymentButtons' ) {
+			return this.state.selectedPaymentId;
+		}
+
+		if ( activeTab === 'addNew' ) {
+			// ask the form controller to validate the field values
+			const hasErrors = await new Promise( resolve =>
+				this.formStateController.handleSubmit( resolve ),
+			);
+
+			if ( hasErrors ) {
+				return null;
+			}
+
+			// ask the data layer to add a new product
```

Comment:
note that the UI never asks the data layer to do anything. it's a small nuance, but the UI _only_ concerns itself with UI concerns. the data layer watches for the events flowing by and fulfills data needs.